delta-lake/delta-23x/src/main/scala/org/apache/spark/sql/delta/rapids/delta23x/GpuUpdateCommand.scala

```diff
@@ -0,0 +1,275 @@
+    // the BasicWriteStatsTracker. This is for case 2 where the update condition contains only
+    // metadata predicates and so the entire partition is re-written.
+    val outputRows = txn.getMetric("numOutputRows").map(_.value).getOrElse(-1L)
+    if (metrics("numUpdatedRows").value == 0 && outputRows != 0 &&
+      metrics("numCopiedRows").value == 0) {
+      // We know that numTouchedRows = numCopiedRows + numUpdatedRows.
+      // Since an entire partition was re-written, no rows were copied.
+      // So numTouchedRows == numUpdateRows
+      metrics("numUpdatedRows").set(metrics("numTouchedRows").value)
+    } else {
+      // This is for case 3 where the update condition contains both metadata and data predicates
+      // so relevant files will have some rows updated and some rows copied. We don't need to
+      // consider case 1 here, where no files match the update condition, as we know that
+      // `totalActions` is empty.
+      metrics("numCopiedRows").set(
+        metrics("numTouchedRows").value - metrics("numUpdatedRows").value)
+    }
+    txn.registerSQLMetrics(sparkSession, metrics)
+    val finalActions = createSetTransaction(sparkSession, deltaLog).toSeq ++ totalActions
+    txn.commitIfNeeded(finalActions, DeltaOperations.Update(condition.map(_.toString)))
+    sendDriverMetrics(sparkSession, metrics)
+
+    recordDeltaEvent(
+      deltaLog,
+      "delta.dml.update.stats",
+      data = UpdateMetric(
+        condition = condition.map(_.sql).getOrElse("true"),
+        numFilesTotal,
+        numTouchedFiles,
+        numRewrittenFiles,
+        numAddedChangeFiles,
+        changeFileBytes,
+        scanTimeMs,
+        rewriteTimeMs)
+    )
+  }
+
+  /**
+   * Scan all the affected files and write out the updated files.
+   *
+   * When CDF is enabled, includes the generation of CDC preimage and postimage columns for
+   * changed rows.
+   *
+   * @return the list of [[AddFile]]s and [[AddCDCFile]]s that have been written.
+   */
+  private def rewriteFiles(
+      spark: SparkSession,
+      txn: OptimisticTransaction,
+      rootPath: Path,
+      inputLeafFiles: Seq[String],
+      nameToAddFileMap: Map[String, AddFile],
+      condition: Expression): Seq[FileAction] = {
+    // Containing the map from the relative file path to AddFile
+    val baseRelation = buildBaseRelation(
+      spark, txn, "update", rootPath, inputLeafFiles, nameToAddFileMap)
+    val newTarget = DeltaTableUtils.replaceFileIndex(target, baseRelation.location)
+    val targetDf = Dataset.ofRows(spark, newTarget)
+
+    // Number of total rows that we have seen, i.e. are either copying or updating (sum of both).
+    // This will be used later, along with numUpdatedRows, to determine numCopiedRows.
+    val numTouchedRows = metrics("numTouchedRows")
+    val numTouchedRowsUdf = DeltaUDF.boolean {
+      new GpuDeltaMetricUpdateUDF(numTouchedRows)
+    }.asNondeterministic()
```

Comment:
Same here do we want this deterministic?

Reply:
Same case here, it's only used in a filter and matches CPU behavior.